Primary Menu: Fix alignment and scrolling on mobile

ryelle opened the pull request:
Fixes #189 — The menu/close button now stays fixed at the top of the screen, while long menu items can be scrolled through. Menu items move under the button container, so the text doesn't overlap anymore. This also fixes the content shift described in #216. 

|           | Logged in | Not logged in |
|-----------|-----------|---------------|
| initial   | ![Screen Shot 2020-10-05 at 8 20 21 PM](https://user-images.githubusercontent.com/541093/95145113-5a5cc100-0748-11eb-8d4f-d2e526d69e65.png) | ![Screen Shot 2020-10-05 at 8 16 21 PM](https://user-images.githubusercontent.com/541093/95144939-ec17fe80-0747-11eb-9106-62afe1336990.png) |
| menu open | ![Screen Shot 2020-10-05 at 8 20 55 PM](https://user-images.githubusercontent.com/541093/95145107-54ff7680-0748-11eb-85d1-990504479b78.png) | ![Screen Shot 2020-10-05 at 8 16 30 PM](https://user-images.githubusercontent.com/541093/95144968-ffc36500-0747-11eb-84f6-753f058e4bef.png) |
| scrolled  | ![Screen Shot 2020-10-05 at 8 21 02 PM](https://user-images.githubusercontent.com/541093/95145101-4e70ff00-0748-11eb-9c11-55cb59604d5f.png) | ![Screen Shot 2020-10-05 at 8 17 52 PM](https://user-images.githubusercontent.com/541093/95144978-05b94600-0748-11eb-9243-5b2ba642aec1.png) |

⚠️  I couldn't get my test site to work in Firefox or Safari, so this has _only been tested in Chrome/Mac._  I'll try to fix my setup, but until then other folks testing in other browsers or actually on mobile devices would be a big help.

ryelle committed:
Primary Menu: Fix alignment and scrolling on mobile

ryelle committed:
Style header to match smaller-screen mockup

carolinan commented:
I still see a visible shift where the content moves to the left when the button is clicked.
In this screenshot, I have removed the background color of the menu just to confirm:

![The content below the menu shifts to the left](https://user-images.githubusercontent.com/7422055/95161958-c40ab880-07a4-11eb-8e6b-de50f2fc9f24.png)

I am not sure if the devtools are affecting the menu and keyboard testing here in any way. It makes me wish that the responsive menu appeared on a wider width.

The scroll works well for me.
In my tests, a part of the focus outline on the first menu item is cut off, and it is cut off in various degrees in different browsers:

Chrome
![ menu item focus outline is cut off in Chrome ](https://user-images.githubusercontent.com/7422055/95165794-020bda80-07ad-11eb-8100-1fcb8ff13220.png)

Firefox:
![ menu item focus outline is cut off in Firefox](https://user-images.githubusercontent.com/7422055/95165900-37b0c380-07ad-11eb-88d6-ff3c40ebc778.png)

carolinan commented on assets/sass/06-components/navigation.scss:
```diff
@@ -89,34 +88,41 @@
 			z-index: 499;
 		}
+
+		.admin-bar & {
```
This is affecting the wide menu as well.

carolinan commented on assets/sass/06-components/navigation.scss:
```diff
@@ -89,34 +88,41 @@
 		}
+
+		.admin-bar & {
+			height: calc(100vh - 46px);
```
This is height is affecting the wide menu as well.

ryelle replied:
ah, thanks for catching that. Fixed in https://github.com/WordPress/twentytwentyone/pull/286/commits/32c420deb097acd8ea1b26db43c7c6b3ae5336ff

carolinan commented:
In Firefox on PC, an extra tab is needed to reach the first menu item.
I do not see it in any other browsers,  and not without this PR applied.

1) Tab to the menu button and open the menu
2) Press tab: - Focus is lost or is on an item that is not visible.
3) Press tab again: Focus is on the first menu item.

The same thing happens when you tab backwards between the first menu item and the close button.

aristath commented:
> In Firefox on PC, an extra tab is needed to reach the first menu item.

I can confirm this (Firefox on Ubuntu)

ryelle committed:
Only set the nav height on mobile sizes

ryelle committed:
Fix the horizontal shift of content as the menu opens

ryelle committed:
Fix the menu offset so the top & bottom items aren't cut off

ryelle committed:
Add build files

ryelle commented:
I've fixed the menu alignment so the top focus outline should not be cut off anymore. I had also misunderstood the alignment issue (I wasn't watching the disappearing content), but the transparent menu made that clear — so that's also fixed in https://github.com/WordPress/twentytwentyone/pull/286/commits/8a4eaea412657e05cad11e1f10440b9698eeb5d7.

carolinan commented:
Looks good to me, I feel that this can be merged, I am no closer to finding why Firefox has an extra tab, perhaps it is JS related.

ryelle commented:
Good news: I fixed my local environment & was able to test in Firefox
Bad news: The tab stop issue exists on trunk too, and appears to be [behavior specific to Firefox](https://bugzilla.mozilla.org/show_bug.cgi?id=616594) when there's `overflow: auto` with scrolling content:

> it's the only way for a keyboard-only user to scroll it (by tabbing to it and then using arrow keys and the like).

(The element focused is `primary-menu-container`)

We might want to add an outline style to the menu just to make the active element more obvious, but that can be another issue.

carolinan commented:
what about reverting the overflow:auto?

ryelle commented:
It looks like this happens with any overflow property, which we can't remove because it's needed to allow scrolling with a mouse.

jffng commented:
I tried this out on my mobile device (iPhone 11) and here's what I'm getting:
 
Chrome | Safari
------- | -------
![IMG_AF7AA954D077-1](https://user-images.githubusercontent.com/5375500/95266676-7f087580-07e8-11eb-935f-6f4d5fbafae4.jpeg) | ![IMG_4FDC6A9E6A88-1](https://user-images.githubusercontent.com/5375500/95266709-934c7280-07e8-11eb-8591-959149a3abd4.jpeg)

The menus won't open in either browser — I'm not sure if this is related to my tunnel, but I can upload the PR to a different test site and check.

carolinan commented:
Tested the PR on my Huawei phone in both orientations on Firefox, Chrome, Opera mini, Edge.
On ipad in Chrome, Safari. Edge and Firefox.  The menu and the buttons work as expected. 

I'd like to see it paired with https://github.com/WordPress/twentytwentyone/pull/290, either on trunk or if trunk can be merged back into this PR, if you want to do more testing.

ryelle commented:
I've rebased the branch so it should have that change— in a quick look I don't think there would be any interaction, but if there is I can fix it in (my) morning.

carolinan commented:
Let's merge this and make a note to retest on Iphone.
